### PR TITLE
Step 3: Add Content Edit Features to prevent cursor to be stuck between ZeroWidthSpace & ReadOnly Entity

### PR DESCRIPTION
**Pull request by BryanValverdeU**
Add 3 new Content Edit Features
1. When Right Arrow Key in the before delimiter, move the cursor to the delimiter After
2. When Left Arrow Key in the After Delimiter, move the cursor to the delimiter Before
3. When Backspace or Delete is pressed in the delimiter, Remove the Entity

Delimiter2

### CODE REVIEW

File: packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts [hunk 1]
```diff
@@ -191,6 +196,216 @@ function cacheGetNeighborEntityElement(
     return element;
 }
 
+/**
+ * Content edit feature to move the cursor from Delimiter After Entity to the delimiter Before
+ */
+const MoveBeforeDelimiterFeature: BuildInEditFeature<PluginKeyboardEvent> = {
+    keys: [Keys.LEFT],
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts:203]: Do we need special handling for RTL?

File: packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts [hunk 2]
```diff
@@ -191,6 +196,216 @@ function cacheGetNeighborEntityElement(
...
+    shouldHandleEvent: (event: PluginKeyboardEvent, editor: IEditor) => {
+        return getIsDelimiterAtCursor(event, editor, false /* checkBefore */);
+    },
+    handleEvent(event: PluginKeyboardEvent, editor: IEditor) {
+        const delimiter = cacheDelimiter(event, false /* checkBefore */);
+        if (!delimiter) {
+            return;
+        }
+        const delimiterBefore = delimiter.previousElementSibling?.previousElementSibling;
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts:212]: Do we also need to check previous sibling is entity?

File: packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts [hunk 3]
```diff
@@ -191,6 +196,216 @@ function cacheGetNeighborEntityElement(
...
+
+            if (selection) {
+                event.rawEvent.preventDefault();
+                editor.runAsync(() => {
+                    const position = new Position(delimiterBefore, PositionType.Before);
+                    if (event.rawEvent.shiftKey) {
+                        selection.extend(position.element, position.offset);
+                    } else {
+                        selection.setPosition(position.element, position.offset);
+                    }
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts:225]: You can do editor.select(delimiterBefore, PositionType.Before)

File: packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts [hunk 4]
```diff
@@ -191,6 +196,216 @@ function cacheGetNeighborEntityElement(
...
+    checkBefore: boolean,
+    entity?: Element | null,
+    operation?: EntityOperation
+) {
+    const element = cacheGetEventData(event, 'entity_delimiter_cache_key_' + checkBefore, () =>
+        editor.getElementAtCursor(getEntitySelector(), entity)
+    );
+
+    if (element && operation !== undefined) {
+        editor.triggerPluginEvent(PluginEventType.EntityOperation, {
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts:399]: Do we expect plugin to help delete the delimiters?

File: packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts [hunk 5]
```diff
@@ -191,6 +196,216 @@ function cacheGetNeighborEntityElement(
...
+};
+
+/**
+ * Content edit Feature to trigger a Delete Entity Operation when one of the Delimiter is about to be removed with DELETE or Backspace
+ */
+const RemoveEntityBetweenDelimitersFeature: BuildInEditFeature<PluginKeyboardEvent> = {
+    keys: [Keys.BACKSPACE, Keys.DELETE],
+    shouldHandleEvent(event: PluginKeyboardEvent, editor: IEditor) {
+        const range = editor.getSelectionRange();
+        if (!range.collapsed) {
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts:274]: If press SHIFT+Left/Right, I think we should also handle the selection. For example, cursor is after delimiter, press SHIFT+LEFT, need to select the whole entity as well as two delimiters.

File: packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts (hunk 1 above)
Comment (BryanValverdeU) [reply packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts:203]: Ah I see, yes we need to do that.

File: packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts (hunk 4 above)
Comment (BryanValverdeU) [reply packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts:399]: I added some logic in previous PR to delete the delimeters when a entity is deleted [here](https://github.com/microsoft/roosterjs/blob/0341fa88ecb018f2f09964ee87249f6848c94a31/packages/roosterjs-editor-core/lib/corePlugins/utils/InlineEntityHandlers/inlineEntityOnPluginEvent.ts#L41). Not sure if that is okay or we should move the logic elsewhere

File: packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts (hunk 1 above)
Comment (BryanValverdeU) [reply packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts:203]: Should be fixed in latest commit
Delimiter2

File: packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts (hunk 2 above)
Comment (BryanValverdeU) [reply packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts:212]: Good point, added a check too.

File: packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts (hunk 5 above)
Comment (BryanValverdeU) [reply packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts:274]: Added in the latest commit:
Delimiter3

File: packages/roosterjs-editor-types/lib/interface/ContentEditFeatureSettings.ts [hunk 6]
```diff
@@ -62,6 +62,16 @@ export interface EntityFeatureSettings {
      * press DELETE right after an entity
      */
     deleteBeforeEntity: boolean;
+
+    /**
+     * Content edit feature to move the cursor from Delimiters around Entities when using Right or Left Arrow Keys
+     */
+    MoveBetweenDelimitersFeature: boolean;
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-types/lib/interface/ContentEditFeatureSettings.ts:69]: start with lower case

File: packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts [hunk 7]
```diff
@@ -191,6 +198,227 @@ function cacheGetNeighborEntityElement(
...
+                shouldCheckBefore = true;
+            }
+        }
+        if (event.rawEvent.which === Keys.LEFT) {
+            if (isRTL) {
+                shouldCheckBefore = true;
+            } else {
+                shouldCheckBefore = false;
+            }
+        }
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/entityFeatures.ts:223]: How about: 
const shouldCheckBefore = isRTL == (event.rawEvent.which === Keys.LEFT);

**Review by JiuqingSong — APPROVED** [review]

**Pull request merged**